ui/appui-layout-react/src/appui-layout-react/target/useAllowedWidgetTarget.tsx

```diff
@@ -0,0 +1,60 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Widget
+ */
+import * as React from "react";
+import { DraggedWidgetIdContext } from "../base/DragManager";
+import { DraggedTabStateContext, NineZoneContext, TabsStateContext, WidgetsStateContext } from "../base/NineZone";
+import { getWidgetLocation, isFloatingWidgetLocation, isPopoutWidgetLocation } from "../state/WidgetLocation";
+import { PanelSide } from "../widget-panels/Panel";
+import { WidgetState } from "../state/WidgetState";
+import { assert } from "@itwin/core-bentley";
+
+/** Checks the proposed docking target to see if it's allowed by the dragged widget or tab
+ * @internal
+ */
+export function useAllowedWidgetTarget(widgetId: WidgetState["id"]) {
```

Comment:
Would be great to have an internal helper for code reuse in `useAllowedWidgetTarget` and `useAllowedPanelTarget` hooks..

Reply:
Agreed. I couldn't get one coded that didn't violate the no conditional hooks rule, so I left it as is.

Reply:
From what I see, useAllowedPanelTarget.tsx lines 20 to 46 could be extracted as an internal hook (say useAllowedSideTarget) that would only have "side" as input, 

```
export function useAllowedPanelTarget() {
  const side = React.useContext(PanelSideContext);
  return useAllowedSideTarget(side);
}

export function useAllowedWidgetTarget(widgetId: WidgetState["id"]) {
  const state ...
  const widgetLocation ...
  if(!widgetLocation...) return false;
  if(isFloating...) return true;
  return useAllowedSideTarget(widgetLocation.side);
}
```

Reply:
That implementation of useAllowedWidgetTarget() calls useAllowedSideTarget() conditionally.

Reply:
Right indeed!
Ok, suggestion 2 then:
```
function useAllowedSideTarget(side: PanelSide, knownReturn: boolean | undefined) {
  const draggedTab = React.useContext(DraggedTabStateContext);
  const draggedWidget = React.useContext(DraggedWidgetIdContext);
  const tabsState = React.useContext(TabsStateContext);
  const widgetsState = React.useContext(WidgetsStateContext);
  if(knownReturn !== undefined) {
    return knownReturn;
  }
  let allowedPanelTargets...
  ...
}

export function useAllowedWidgetTarget(widgetId: WidgetState["id"]) {
  ...
  let knownResult: boolean | undefined;
  if(!widgetLocation) knownResult = false;
  else if(isFloating) knownResult = true;
  return useAllowedSideTarget(side, knownResult);
}
```
It will call unneeded context, but it called it anyway...

Reply:
Done!